specifyweb/specify/tree_views.py

```diff
@@ -343,4 +351,392 @@ def perm_target(tree):
+    }[tree]
+
+TAXON_RANKS = {
+    'TAXONOMY_ROOT': 0,
+    'TAXONOMY ROOT': 0,
+    'LIFE': 0,
+    'KINGDOM': 10,
+    'SUBKINGDOM': 20,
+    'DIVISION': 30,
+    'PHYLUM': 30,
+    'SUBDIVISION': 40,
+    'SUBPHYLUM': 40,
+    'SUPERCLASS': 50,
+    'CLASS': 60,
+    'SUBCLASS': 70,
+    'INFRACLASS': 80,
+    'SUPERORDER': 90,
+    'ORDER': 100,
+    'SUBORDER': 110,
+    'INFRAORDER': 120,
+    'PARVORDER': 125,
+    'SUPERFAMILY': 130,
+    'FAMILY': 140,
+    'SUBFAMILY': 150,
+    'TRIBE': 160,
+    'SUBTRIBE': 170,
+    'GENUS': 180,
+    'SUBGENUS': 190,
+    'SECTION': 200,
+    'SUBSECTION': 210,
+    'SPECIES': 220,
+    'SUBSPECIES': 230,
+    'VARIETY': 240,
+    'SUBVARIETY': 250,
+    'FORMA': 260,
+    'SUBFORMA': 270
+}
+GEOGRAPHY_RANKS = {
+    'CONTINENT': 100,
+    'COUNTRY': 200,
+    'STATE': 300,
+    'COUNTY': 400
+}
+STORAGE_RANKS = {
+    'BUILDING': 100,
+    'COLLECTION': 150,
+    'ROOM': 200,
+    'AISLE': 250,
+    'CABINET': 300,
+    'SHELF': 350,
+    'BOX': 400,
+    'RACK': 450,
+    'VIAL': 500
+}
+GEOLOGIC_TIME_PERIOD_RANKS = {
+    'ERA': 100,
+    'PERIOD': 200,
+    'EPOCH': 300,
+    'AGE': 400
+}
+LITHO_STRAT_RANKS = {
+    'SUPERGROUP': 100,
+    'GROUP': 200,
+    'FORMATION': 300,
+    'MEMBER': 400,
+    'BED': 500
+}
+
+DEFAULT_RANK_INCREMENT = 100
+TAXON_RANK_INCREMENT = 10
+GEOGRAPHY_RANK_INCREMENT = DEFAULT_RANK_INCREMENT
+STORAGE_RANK_INCREMENT = 50
+GEOLOGIC_TIME_PERIOD_INCREMENT = DEFAULT_RANK_INCREMENT
+LITHO_STRAT_INCREMENT = DEFAULT_RANK_INCREMENT
+
+# Map tree type to default tree ranks and default rank id increment
+TREE_RANKS_MAPPING = {
+    'taxon': (TAXON_RANKS, TAXON_RANK_INCREMENT),
+    'geography': (GEOGRAPHY_RANKS, GEOGRAPHY_RANK_INCREMENT),
+    'storage': (STORAGE_RANKS, STORAGE_RANK_INCREMENT),
+    'geologictimeperiod': (GEOLOGIC_TIME_PERIOD_RANKS, GEOLOGIC_TIME_PERIOD_INCREMENT),
+    'lithostrat': (LITHO_STRAT_RANKS, LITHO_STRAT_INCREMENT),
+}
+
+# Wrapper for tree editing api functions
+def tree_edit(edit_func):
+    @login_maybe_required
+    @require_POST
+    @transaction.atomic
+    @wraps(edit_func)
+    def wrapper(*args, **kwargs):
+        try:
+            edit_func(*args, **kwargs)
+            result = {'success': True}
+        except BusinessRuleException as e:
+            result = {'success': False, 'error': str(e)}
+            return HttpResponseServerError(toJson(result), content_type="application/json")
+        return HttpResponse(toJson(result), content_type="application/json")
+
+    return wrapper
+
+@openapi(schema={
+    'post': {
+        "requestBody": {
+            "required": True,
+            "description": "Add rank to a tree.",
+            "content": {
+                "application/json": {
+                    "schema": {
+                        "type": "object",
+                        "description": "The request body.",
+                        "properties": {
+                            "newRankName": {
```

Comment:
instead of coming up with new arguments, could we just let front-end pass the entire <TreeName>TreeDefItem resource to this endpoint?

OR EVEN BETTER, just have front-end call the regular /api/specify/<TreeName>TreeDefItem endpoint, and then only call this endpoint with an id of the created rank id and the desired parent rank - that way, you don't duplicate code as a lot of the work would be handled by the /api/specify/<TreeName>TreeDefItem endpoint.

at which point, pretty much the only thing this endpoint would be doing is setting corect rank id and regenerating full_names - which is pretty much exactly what a `move_tree_rank` endpoint would look like, so we just got a feature for free, while reducing code duplication!

Reply:
oh, I just realized that a `parent` relationship exists in the TaxonTreeDefItem table - so this endpoint shouldn't exist at all, and the work should happen in the pre_save and post_save `orm_signal_handler`

the goal is not to add API endpoints with gazillion options, but to make it "just work" with the smallest API possible, to make it easier for anyone using the API

also, having these actions happen automatically in pre/post save/delete hook creates a "pit of success", where even if you don't know that special cleanup logic is needed for tree ranks, or are editing tree ranks directly though a Specify Form, rather than inside the Tree View, things would still "just work"